jscomp/others/js_dict.ml

```diff
@@ -24,29 +24,64 @@
 
 (** Provides a simple key-value dictionary abstraction over native JavaScript objects *)
 
-(** The dict type *)
-type 'a t
 
-(** The key type, an alias of string *)
+type 'a t
 type key = string
 
-(** [get dict key] returns the value associated with [key] in [dict] *)
-external get : 'a t -> key -> 'a option = "" [@@bs.get_index] [@@bs.return {undefined_to_opt}]
-
-(** [unsafeGet dict key] returns the value associated with [key] in [dict]
+external empty : unit -> 'a t = "" [@@bs.obj]
 
-This function will return an invalid value ([undefined]) if [key] does not exist in [dict]. It
-will not throw an error.
-*)
+external get : 'a t -> key -> 'a option = "" [@@bs.get_index] [@@bs.return {undefined_to_opt}]
 external unsafeGet : 'a t -> key -> 'a = "" [@@bs.get_index] 
-
-(** [set dict key value] sets the value of [key] in [dict] to [value] *)
 external set : 'a t -> key -> 'a -> unit = "" [@@bs.set_index]  
-
-(** [keys dict] returns an array of all the keys in [dict] *)
 external keys : 'a t -> key array = "Object.keys" [@@bs.val]
 
-(** [empty ()] creates an empty dictionary *)
-external empty : unit -> 'a t = "" [@@bs.obj]
+(* external entries : 'a t -> (key * 'a) array = "Object.entries" [@@bs.val] (* ES2017 *) *)
+let entries dict =
+  let keys = keys dict in
+  let l = Js.Array.length keys in
+  let values = Obj.magic (Array.make l 0) in
+  for i = 0 to l - 1 do
+    let key = keys.(i) in
+    values.(i) <- (key, unsafeGet dict key)
+  done;
+  values
+
+(* external values : 'a t -> 'a array = "Object.values" [@@bs.val] (* ES2017 *) *)
+let values dict =
+  let keys = keys dict in
+  let l = Js.Array.length keys in
+  let values = Obj.magic (Array.make l 0) in
+  for i = 0 to l - 1 do
+    values.(i) <- unsafeGet dict keys.(i)
+  done;
+  values
+
+let fromList entries =
+  let dict = empty () in
+  let rec loop = function
+  | [] -> dict
+  | (key, value) :: rest ->
+    set dict key value;
+    loop rest
+  in
+  loop entries
+
+let fromArray entries =
+  let dict = empty () in
+  let l = Js_array.length entries in
+  for i = 0 to l - 1 do
+    let (key, value) = entries.(i) in
```

Comment:
I thought we were thought we were trying to reduce the dependency on external modules, and that these were equivalent. But just so I understand the difference: `.(i)` will do bounds checking and raise, while ` unsafe_get` would just return whatever if out of bounds?